command/metadata.go

```diff
@@ -0,0 +1,136 @@
+package command
+
+import (
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/mackerelio/mackerel-agent/config"
+	"github.com/mackerelio/mackerel-agent/metadata"
+)
+
+func metadataGenerators(conf *config.Config) []*metadata.Generator {
+	generators := make([]*metadata.Generator, 0, len(conf.MetadataPlugins))
+
+	workdir := os.Getenv("MACKEREL_PLUGIN_WORKDIR")
+	if workdir == "" {
+		workdir = os.TempDir()
+	}
+
+	for name, pluginConfig := range conf.MetadataPlugins {
+		generator := &metadata.Generator{
+			Name:     name,
+			Config:   pluginConfig,
+			Tempfile: filepath.Join(workdir, "mackerel-metadata", name),
+		}
+		logger.Debugf("Metadata plugin generator created: %#v %#v", generator, generator.Config)
+		generators = append(generators, generator)
+	}
+
+	return generators
+}
+
+type metadataResult struct {
+	namespace string
+	metadata  interface{}
+}
+
+func runMetadataLoop(c *Context, termMetadataCh <-chan struct{}, quit <-chan struct{}) {
+	resultCh := make(chan *metadataResult)
+	for _, g := range c.Agent.MetadataGenerators {
+		go runEachMetadataLoop(g, resultCh, quit)
+	}
+
+	exit := false
+	for !exit {
+		select {
+		case <-time.After(1 * time.Minute):
+		case <-termMetadataCh:
+			logger.Debugf("received 'term' chan for metadata loop")
+			exit = true
+		}
+
+		results := make(map[string]*metadataResult)
+	ConsumeResults:
+		for {
+			select {
+			case result := <-resultCh:
+				// prefer new result to old metadata which remains to be posted on retry
+				// and also avoid infinite number of retries
+				results[result.namespace] = result
+			default:
+				break ConsumeResults
+			}
+		}
+
+		for _, result := range results {
+			resp, err := c.API.PutMetadata(c.Host.ID, result.namespace, result.metadata)
+			// retry on 5XX errors
+			if resp != nil && resp.StatusCode >= 500 {
+				logger.Errorf("put metadata %q failed: status %s", result.namespace, resp.Status)
+				resultCh <- &metadataResult{
```

Comment:
You should use goroutine here. Otherwise it causes a deadlock. Because it will send and receive to one channel within the same thread.